apps/muddle-node/main.cpp

```diff
@@ -0,0 +1,300 @@
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "core/commandline/parameter_parser.hpp"
+#include "core/containers/set_difference.hpp"
+#include "core/filesystem/read_file_contents.hpp"
+#include "core/filesystem/write_to_file.hpp"
+#include "core/macros.hpp"
+#include "core/threading/synchronised_state.hpp"
+#include "crypto/ecdsa.hpp"
+#include "muddle/muddle_endpoint.hpp"
+#include "muddle/muddle_interface.hpp"
+#include "network/management/network_manager.hpp"
+#include "http/module.hpp"
+#include "http/server.hpp"
+#include "telemetry/registry.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <csignal>
+#include <cstdlib>
+#include <list>
+#include <random>
+#include <thread>
+
+using namespace std::this_thread;
+using namespace std::chrono;
+using namespace std::chrono_literals;
+using namespace fetch;
+using namespace fetch::network;
+using namespace fetch::muddle;
+
+namespace {
+
+using fetch::commandline::ParamsParser;
+
+using AddressList   = std::list<Address>;
+using AddressSet    = std::unordered_set<Address>;
+using MsgCounters   = std::unordered_map<Address, std::size_t>;
+using Clock         = std::chrono::steady_clock;
+using Timepoint     = Clock::time_point;
+using HttpServerPtr = std::unique_ptr<http::HTTPServer>;
+using HttpModulePtr = std::unique_ptr<http::HTTPModule>;
+
+struct AggregateData
+{
+  MsgCounters counters{};
+  std::size_t total_messages{0};
+};
+
+using Statistics = fetch::SynchronisedState<AggregateData>;
+
+std::atomic<bool>        gActive{true};
+std::atomic<std::size_t> gInterruptCount{0};
+
+Statistics gStatistics;
+
+constexpr uint16_t    SERVICE{1};
+constexpr uint16_t    CHANNEL{1};
+constexpr char const *LOGGING_NAME{"main"};
+
+/**
+ * The main interrupt handler for the application
+ */
+void InterruptHandler(int /*signal*/)
+{
+  std::size_t const interrupt_count = ++gInterruptCount;
+
+  if (interrupt_count > 1)
+  {
+    FETCH_LOG_INFO(LOGGING_NAME, "User requests stop of service (count: ", interrupt_count, ")");
+  }
+  else
+  {
+    FETCH_LOG_INFO(LOGGING_NAME, "User requests stop of service");
+  }
+
+  // signal that the program should stop
+  gActive = false;
+
+  if (interrupt_count >= 3)
+  {
+    std::exit(1);
+  }
+}
+
+ProverPtr RestoreOrCreateKey(ParamsParser const &params)
```

Comment:
It feels like this is something that would be useful to have some where in the lib itself such that other apps can use it.

Reply:
Good suggestion -  I have another update in the pipeline will add it to the list